### ISSUE TITLE
Generic type of superclass cannot be inferred from function call on a subclass

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/TypeScript#20348]
**TypeScript Version:**  2.7.0-dev.20171129

**Code**

```ts
class Evented<M> {
    // protected __typeMap__?: M; 
    on<K extends keyof M>(type: K, listener: (event: M[K]) => void) { }
}

function on<M, K extends keyof M>(target: Evented<M>, type: K, listener: (event: M[K]) => void) { }

interface EventMap {
    change: { type: 'change'; };
}

const e = new Evented<EventMap>();
e.on('change', (event) => { }); // type of event: { type: 'change' }
on(e, 'change', (event) => { }); // type of event: { type: 'change' }

class SubEvented extends Evented<EventMap> { }

const s = new SubEvented();
s.on('change', (event) => { }); // type of event: { type: 'change' }
on(s, 'change', (event) => { }); // type of event: M[K]
```

**Expected behavior:**
`event` is inferred as `{ type: 'change' }` on line 20 (last line)

**Actual behavior:**
Line 20 is flagged as an error:

```
error TS2345: Argument of type '"change"' is not assignable to parameter of type 'never'.
```

Currently, the only way to get line 20 to work as expected is to uncomment line 2.

**Comment by mhegazy** [comment]
Should be fixed by https://github.com/Microsoft/TypeScript/pull/20126

**Comment by mhegazy** [comment]
it is worth noting that the inference here  for `M` will not be `EventMap`, since the there is no manifestation for this type to be inferred from. however, with #20126, `M` will be inferred from the use `M[K]` to be a type with a property named `K` (or "change" in the example above).

**Comment by RyanCavanaugh** [comment]
This is the intended behavior -- there is no inference site for `M`
